shopfloor_reception/services/reception.py

```diff
@@ -0,0 +1,877 @@
+                    lot_name, products=selected_lines.mapped("product_id")
+                )
+                if not lot:
+                    lot = lot_model.create(
+                        {
+                            "name": lot_name,
+                            "product_id": line.product_id.id,
+                            "company_id": self.env.company.id,
+                            "use_expiration_date": line.product_id.use_expiration_date,
+                        }
+                    )
+                line.lot_id = lot.id
+                # TODO: Dunno why, I have to explicitely call the onchange method
+                line._onchange_lot_id()
+            return self._response_for_set_lot(picking, selected_lines)
+        elif expiration_date:
+            selected_lines.write({"expiration_date": expiration_date})
+            selected_lines.mapped("lot_id").write({"expiration_date": expiration_date})
+            return self._response_for_set_lot(picking, selected_lines)
+        return self._response_for_set_lot(picking, selected_lines)
+
+    def set_lot_confirm_action(self, picking_id, selected_line_ids):
+        picking = self.env["stock.picking"].browse(picking_id)
+        selected_lines = self.env["stock.move.line"].browse(selected_line_ids)
+        message = self._check_expiry_date(picking)
+        if message:
+            return self._response_for_set_lot(picking, selected_lines, message=message)
+        return self._response_for_set_quantity(picking, selected_lines)
+
+    def _check_expiry_date(self, picking):
+        for line in picking.move_line_ids:
+            use_expiration_date = (
+                line.product_id.use_expiration_date or line.lot_id.use_expiration_date
+            )
+            if use_expiration_date and not line.expiration_date:
+                return {
+                    "message_type": "error",
+                    "body": _("Missing expiration date."),
+                }
+
+    def set_quantity(
+        self,
+        picking_id,
+        selected_line_ids,
+        quantity=None,
+        barcode=None,
+        confirmation=False,
+    ):
+        """Set the quantity done
+
+        Input:
+            quantity: the quantity to set
+            barcode: Barcode of a product / packaging to determine the qty to increment
+            barcode: Barcode of a package / location to set on the line
+
+        transitions:
+          - select_line: User clicked on back
+          - set_lot: Barcode not found. Ask user to create one from barcode
+          - set_lot: expiration_date has been set on the selected line
+          - set_lot: lot_it has been set on the selected line
+          - set_lot: Error: expiration_date is required
+          - set_quantity: User clicked on the confirm button
+        """
+        picking = self.env["stock.picking"].browse(picking_id)
+        selected_lines = self.env["stock.move.line"].browse(selected_line_ids)
```

Comment:
Do not hesitate to check if the ID of the records you are receiving really exist (with `exists()` method), like here: https://github.com/OCA/wms/blob/14.0/shopfloor/services/zone_picking.py#L768-L770